client/my-sites/stats/hooks/use-wpcom-paid-stats.ts

```diff
@@ -0,0 +1,59 @@
+import { isEnabled } from '@automattic/calypso-config';
+import { FEATURE_STATS_PAID } from '@automattic/calypso-products';
+import { useSelector } from 'calypso/state';
+import siteHasFeature from 'calypso/state/selectors/site-has-feature';
+import { isJetpackSite } from 'calypso/state/sites/selectors';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+
+// TODO: define shared variables
+const trafficPaidStats = [
+	'statsSearchTerms',
+	'statsClicks',
+	'statsReferrers',
+	'statsCountryViews',
+];
+
+const featureFlags = [ 'stats/date-control' ];
+
+/*
+ * Utility function intended to be used with useSelector or redux connect mapStateToProps.
+ * For example in mapStateToProps:
+ * const hasPaidStatsFeature = hasPaidStatsFeature( state, siteId, 'statsSearchTerms' );
+ */
+export const checkPaidStatsFeature = (
+	state: object,
+	siteId: number | null,
+	featureType: string
+) => {
+	const isPaidStatsEnabled = isEnabled( 'stats/paid-wpcom-v2' );
+	const isOdysseyStats = isEnabled( 'is_running_in_jetpack_site' );
+	const jetpackSite = isJetpackSite( state, siteId );
+	const siteHasPaidStats = siteHasFeature( state, siteId, FEATURE_STATS_PAID );
+
+	// check feature flags
+	if ( ! isPaidStatsEnabled || isOdysseyStats ) {
+		return false;
```

Comment:
I don't know. I feel like if it's the sites that the strategy doesn't want to gate, then should we return `true` instead of `false`? Or a `null` to identify that it's not applicable?

Reply:
Should we still be checking for OdysseyStats, I'm a bit confused by this comment.

Reply:
Needs to return true here to disable the paywall gating when the config flag is not set or it's an odyssey context. Currently this is paywalling my jetpack test site with https://github.com/Automattic/wp-calypso/pull/85126/ applied.

This says you don't have access to a feature (false value) if you're using Jetpack or paid stats is disabled.

Maybe we should name it something like "shouldPaywall" or "shouldGateFeature"?

Reply:
Decided to change the names to not make things confusing, WIP